### ISSUE TITLE
XLA: Help understanding compute path for HLO graph

### ISSUE THREAD

**Issue by lwogulis** [tensorflow/tensorflow#9194]
@aidan-plenert-macdonald and I are trying to figure out how the HLO graph is being passed around during xla computations. I'm moving computations over to a new device and am attempting to intercept the graph to see how it's represented, in order to replicate the structure.

A list of the files where I added print statements is included below. Almost none of them print (only Registrar initialization in hlo_graph_dumper.cc, the Transfer functions in client.cc and service.cc, and the CpuCompiler and XpuCompiler initializations), and I've added statements in almost every function in the attached files.

Where is the HLO graph being assembled and dumped? How do I access the HLO graph?

Note: the 'xpu' folder/files are for my new device - they're replicas of the 'cpu' folder/files in compiler/xla/service, with all mentions of 'cpu' changed to 'xpu'. List of files with print statements:
```
./compiler/aot/compile.cc
./compiler/jit/encapsulate_subgraphs_pass.cc
./compiler/jit/mark_for_compilation_pass.cc
./compiler/tf2xla/kernels/batch_matmul_op.cc
./compiler/tf2xla/kernels/gather_op.cc
./compiler/tf2xla/xla_compiler.cc
./compiler/xla/client/client.cc
./compiler/xla/service/cpu/cpu_compiler.cc
./compiler/xla/service/hlo_computation.cc
./compiler/xla/service/hlo_graph_dumper.cc
./compiler/xla/service/layout_assignment.cc
./compiler/xla/service/service.cc
./compiler/xla/service/user_computation.cc
./compiler/xla/service/xpu/xpu_compiler.cc
./compiler/xla/service/xpu/xpu_executable.cc
./compiler/xla/tests/hlo_test_base.cc
./core/common_runtime/function.cc
./core/common_runtime/graph_optimizer.cc
```

**Comment by asimshankar** [comment]
You might be able to trace the HLO graph dumping from the handling of the [`TF_XLA_FLAGS`](https://github.com/tensorflow/tensorflow/blob/f94e20f2c704f037a113d350fd7c431e3fd75df8/tensorflow/compiler/xla/legacy_flags/parse_flags_from_env.h#L20) environment variable and the [`xla_dump_*`](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/legacy_flags/service_flags.cc#L75) command-line flags.

As XLA is in early stages, documentation (and indeed the implementation) of these details is not something that is finalized and thus easily accessible yet. 

Does that help?

(CC @tatatodd @eliben in case they have something to add)

**Comment by lwogulis** [comment]
I'm running with TF_XLA_FLAGS=--xla_generate_hlo_graph=.* normally. I just ran with TF_XLA_FLAGS=--xla_dump_* (if that's what you meant) and nothing changed

**Comment by hawkinsp** [comment]
We have a not terribly well publicized Google group for developers working on the XLA implementation, you may wish to post there:
https://groups.google.com/forum/#!forum/xla-dev

One possibility is you aren't actually using XLA at all. Can you show the Python code you are attempting to run? How are you enabling XLA?

In general, you can get out HLO graphs by the environment variable TF_XLA_FLAGS=--xla_generate_hlo_graph=.* . 

If you are using the Tensorflow LOG() macros to print things, you also need to make sure logging is enabled. There are two more environment variables TF_CPP_MIN_LOG_LEVEL and TF_CPP_MIN_VLOG_LEVEL that control that.

**Comment by RichDubielzig** [comment]
I was able to reproduce this with the mnist_softmax_xla.py file from the TensorFlow tutorial:

https://raw.githubusercontent.com/tensorflow/tensorflow/master/tensorflow/examples/tutorials/mnist/mnist_softmax_xla.py

I should expect this to break at CpuCompiler::Compile, and have checked via printing that line 63, which sets ON_1, is executed.

``
rdubielzig@swpl-000224:~/tflow$ lldb -- python3 mnist_softmax_xla.py 
(lldb) target create "python3"
Current executable set to 'python3' (x86_64).
(lldb) settings set -- target.run-args  "mnist_softmax_xla.py"
(lldb) b CpuCompiler::Compile
Breakpoint 1: no locations (pending).
WARNING:  Unable to resolve breakpoint to any actual locations.
(lldb) r
Process 10331 launched: '/usr/bin/python3' (x86_64)
2 locations added to breakpoint 1
Process 10331 stopped and restarted: thread 1 received signal: SIGCHLD
Extracting /tmp/tensorflow/mnist/input_data/train-images-idx3-ubyte.gz
Extracting /tmp/tensorflow/mnist/input_data/train-labels-idx1-ubyte.gz
Extracting /tmp/tensorflow/mnist/input_data/t10k-images-idx3-ubyte.gz
Extracting /tmp/tensorflow/mnist/input_data/t10k-labels-idx1-ubyte.gz
2017-04-13 15:35:52.612498: I tensorflow/compiler/xla/service/platform_util.cc:58] platform Host present with 8 visible devices
2017-04-13 15:35:52.616628: I tensorflow/compiler/xla/service/service.cc:183] XLA service 0x2b38780 executing computations on platform Host. Devices:
2017-04-13 15:35:52.616642: I tensorflow/compiler/xla/service/service.cc:191]   StreamExecutor device (0): <undefined>, <undefined>
0.9195
Process 10331 exited with status = 0 (0x00000000)
``

**Comment by lwogulis** [comment]
Thanks! I'll check out the group.

I'm fairly sure I am using XLA, the code I'm running is modified/simplified from the mnist_softmax_xla example:
```
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

import argparse
import sys 

import tensorflow as tf

from tensorflow.examples.tutorials.mnist import input_data
from tensorflow.python.client import timeline

FLAGS = None


def main(_):
    config = tf.ConfigProto(log_device_placement=True)
    jit_level = 0 
    if FLAGS.xla:
        # Turns on XLA JIT compilation.
        jit_level = tf.OptimizerOptions.ON_1
        print('XLA flag on')

    config.graph_options.optimizer_options.global_jit_level = jit_level
    run_metadata = tf.RunMetadata()
    # Creates a session with log_device_placement set to True.
        # Creates a graph.
    with tf.device('/job:localhost/replica:0/task:0/device:XLA_XPU:0'):
        with tf.Session(config=config) as sess:
       # with tf.device('/device:CPU:0'):
            a = tf.constant([1.0, 2.0, 3.0, 4.0, 5.0, 6.0], 
                            shape=[2, 3], name='a')

            b = tf.constant([1.0, 2.0, 3.0, 4.0, 5.0, 6.0], 
                            shape=[3, 2], name='b')
            c = tf.square(tf.tanh(tf.matmul(a, b)))
        # Runs the op.
        print(sess.run(c, 
              options=tf.RunOptions(trace_level=tf.RunOptions.FULL_TRACE),
              run_metadata=run_metadata))
        trace = timeline.Timeline(step_stats=run_metadata.step_stats)
        with open('timeline.ctf.json', 'w') as trace_file:
            trace_file.write(trace.generate_chrome_trace_format())
            trace_file.flush()
        sess.close()

if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument('--data_dir', type=str, 
            default='/tmp/tensorflow/mnist/input_data',
            help='Directory for storing input data')
    parser.add_argument(
              '--xla', type=bool, default=True, help='Turn xla via JIT on')
    FLAGS, unparsed = parser.parse_known_args()
    tf.app.run(main=main, argv=[sys.argv[0]] + unparsed)
```
I also have manually gone into the code and set VLOG_IS_ON to always return true (that seems to deal with the MIN_LOG_LEVEL issues)

**Comment by RichDubielzig** [comment]
The text at the top of the group suggests that the question is better suited for stackoverflow, so that is where I have posted the question:

http://stackoverflow.com/questions/43403089/how-to-exercise-xla-compiler

**Comment by asimshankar** [comment]
@lwogulis : `xla_dump_graph*` was meant to refer to a set of flags, like `xla_dump_hlo_text_to` etc. (see the link in my previous comment: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/legacy_flags/service_flags.cc#L75)

Closing this issue out since it seems that the xla-dev@ mailing list would be a more appropriate forum for XLA (which is still "alpha/experimental") specific questions.

**Comment by hawkinsp** [comment]
Ah! I see your problem. We don't think the CPU JIT is ready for wide use yet, so it does not get enabled globally via OptimizerOptions. (The *GPU* JIT can be enabled via OptimizerOptions.)

Try marking some operators to compile using the "manual" approach described here:
https://www.tensorflow.org/performance/xla/jit

**Comment by yuchen112358** [comment]
Have you succeeded in moving computations over to a new device by XLA? Could you please tell me about your method in detail? @lwogulis

**Comment by aidan-plenert-macdonald** [comment]
@yuchen112358 Unfortunately, neither @lwogulis nor I work at that company anymore. We did manage to get the computation to run on a new "virtual" device (new bindings to the CPU). As far as I am aware, XLA has yet to be mapped to the device. We have normal TF running on the device. @RichDubielzig is the current lead
